### ISSUE TITLE
Oauth error invalid_request: The redirect_uri is missing or not whitelisted

### ISSUE THREAD

**Issue by ajmalafif** [Shopify/shopify_app#151]
I followed the instruction here: http://shopify.github.io/shopify_app/

on localhost (http://localhost.dev) but can't get pass the installation:

<img src="https://www.evernote.com/l/AFnqRJsK_clGaaiu9JsH4BcT4cxFWkMIoaoB/image.png">

I key-in my development store links but it keeps giving me this error:

`Oauth error invalid_request: The redirect_uri is missing or not whitelisted`

<img src="https://www.evernote.com/l/AFmKr91N4FVEmb07PA2jaS1QRew-3L56ekwB/image.png">

I tried all sorts but none works. Any help?

**Comment by raulbrito** [comment]
More people are having this problem (me included). Haven't tried, but somebody claims to have a workaround here: https://ecommerce.shopify.com/c/shopify-apis-and-technology/t/problem-with-embedded-sdk-not-loading-the-iframe-275604

**Comment by kevinhughes27** [comment]
yup this is a bug on our side and a fix is almost ready to go! I am going to close this issue since there is nothing wrong with shopify_app

**Comment by ajmalafif** [comment]
@kevinhughes27 I see thanks for the update. Really appreciate it!

However some feedback I still think that the tutorial/readme can be further detailed like for the shop name do we really just insert `store.myshopify.com` or actually like all the suggested solution in the forum with `redirect_uri`. I'm confused as the instruction is not really clear.

**Comment by kevinhughes27** [comment]
sorry are you talking about what you enter when you install the app (code from this repo) or are you talking about what you enter on the partners page for your api client?

**Comment by ajmalafif** [comment]
oh i meant when i install the app (code from this repo):

<img src="https://www.evernote.com/l/AFnqRJsK_clGaaiu9JsH4BcT4cxFWkMIoaoB/image.png">

and while at it, I always confuse what to enter for partners page a well:

<img src="https://www.evernote.com/l/AFm-JE5ZMFhFWbGgLrdQdle1iMXhmtppj78B/image.png">

**Comment by kevinhughes27** [comment]
what you enter on the partner's page depends on what your apps routes are. Those are the urls Shopify will make requests at during the auth process.

As for the install just enter the url of the shop you want to install the app on. If you get redirected to that shop you entered it right. There is some code that handles adding http etc if you don't add it yourself.

Does that answer your questions?

**Comment by ajmalafif** [comment]
I think yes. That helps clear what should go into those two.

However specifically for `shopify_app` what I did was I clone the repo on local and from there it generates that Install the application page (which I thought that it's like an example app to sandbox with Shopify) on local (say my local url is `localhost.dev`) and use that on Shopify partners.

I inserted the shop url (mystore.myshopify.com) in the app but get 400. The error brought me to the forum posted by @raulbrito above. 

Anyways thanks for being super helpful! I think I'm doing something wrong because I definitely didn't do what the forum peeps suggested. I feel like once I got it I'll do either a video or picture by picture walkthrough to help people in the same position as I am to able to get started.

**Comment by kevinhughes27** [comment]
the forum post is hard to follow - there is currently a bug on our side preventing newly created apps from working. Should be fixed later today

**Comment by ajmalafif** [comment]
I see. I actually tried with old apps too but failed. I guess I should just wait till later today and see how it goes.

Plus I think it's part of i'm being noob and not understanding the basic of things; I build Shopify themes for few Shopify customers so I was thinking of venturing into app development which got me to try things out.

**Comment by kevinhughes27** [comment]
no worries! the auth flow with building any apps these days is frustrating. I think it is kind of the nature of auth that the error messages can't be all that great

**Comment by selvaganeshc1** [comment]
hi @kevinhughes27 ,
 may i know how to resolve " 
What happened?
Oauth error invalid_request: The redirect_uri is missing or not whitelisted
" issue.

**Comment by kevinhughes27** [comment]
The redirect_uri has been added and you need to add it to the initializer when you make a new app. Its in the readme as well, should be pretty straight forward.

**Comment by yinwai92** [comment]
Hi @kevinhughes27 . I tried to insert redirect uri in shopify_app.rb as stated in readme, but when I deploy it gives me this error "NoMethodError: undefined method `redirect_uri=' for #<ShopifyApp::Configuration:0x007f9979b560c0>" . I am clueless about the error, hope that you can help with this. Thank you !

**Comment by kevinhughes27** [comment]
did you update the gem as well? It was a newly added config option

**Comment by yinwai92** [comment]
Hi @kevinhughes27 . Thanks for the reminder. Previously I have yet to update the gem, now I am working with ver. 6.1.3. After installing new gem the redirect_uri option din't give me complain anymore. Thanks!

**Comment by pshoukry** [comment]
I am facing the same issue, I am using version 6.2.0 of shopify_app, and it's a fresh install:

1- created the app in partner dashboard
2- created new app
3- installed this gem
4- used the generator
5- added the same uri in the  redirect_uri of the shopify_app.rb and the app settings
http://localhost:3000/auth/shopify/callback

Getting the same error, is there something I could be missing or i need to check ?

**Comment by pshoukry** [comment]
Found a solution, thanks.

**Comment by kevinhughes27** [comment]
what did you end up needing to do?

**Comment by pshoukry** [comment]
I Faced multiple issues but here is the summary of a complete noob with shopify trying to run:

1. App appends /shop/shopname.myshopify.com to the redirect_uri param in the request of the first call ( so if you have http://localhost:3000/auth/callback get changed to http://localhost:3000/auth/callback/shop/shopname.myshopify.com

sol: you have to manually remove that part from the url

2- in the shopify_app.rb if you change redirect_uri to http://localhost:3000/auth/shopify/callback the setup completes properly but if you choose to run unsafe scripts you get an error 400 oauth not whitelisted if you try to remove the parameter in the top you get a redirect loop

sol: you change the redirect_uri back to "<redirect_uri>" in shopify_app.rb and it works as expected


3- of course any mismatch between the callback in dashboard and that in the redirect_uri parameter in the call gets you the oauthnot whitelisted error ( which is the basic point in debugging

tl;dr make sure both redirect_uri parameter exactly matches the one in the dashboard ( the app appends to it so check manually ) and keep the redirect_uri parameter in shopify_app.rb=  "<redirect_uri>"

**Comment by oleksandriuzikov** [comment]
I'm facing this issue too, seems like the authorize request adds `shop=` parameter to the `redirect_uri` causing redirect_uri to be *not whitelisted*, when I remove the `shop` param from the browser's location bar and hit enter it authenticates successfully.

Can you please give me a hint where to remove the `shop` param in the code?

**Comment by pshoukry** [comment]
1- First time you run the app manually remove it from the address bar in browser
2- in shopify_app.rb leave config.redirect_uri = "\<redirect_uri\>"

Let me know if that fixed your issue.

**Comment by kevinhughes27** [comment]
Can you make a repo reproducing the issue?

**Comment by pshoukry** [comment]
Sure I can, but it's actually happening every time, 
1- create a new rails app ( latest version of rails )
2- Add shopify_app gem and bundle install
3- run the generator 

Then you should see

**First Problem**

1- go to localhost:3000
2- type the name of the shop and click install

Expected behaviour: 
App installs

Actual behaviour:
400 oauth call back not whitelisted error

Workaround
In the address bar go the redirect_uri parameter at its end you will find /shop/shopname.myshopify.com appended to http://localhost:3000/auth/shopify/callback remove it

Installation should work and app appear in the shop.

**Second Problem**

If you change the redirect_uri parameter in shopify_app.rb to http://localhost:3000/auth/shopify/callback

After app installs if you go to shop and navigate to the app ( Allowing insecure scripts ) 

Expected Behaviour
App Loads

Actual behaviour
Error 400 oauth call back not whitelisted again

Workaround
Set redirect_uri parameter in shopify_app.rb to  "\<redirect_uri\>"

**Comment by kevinhughes27** [comment]
Can you submit the repo still because I have made 5 apps in the last 2 days
from scratch exactly how you've described and this has never happened to
me.

**Comment by pshoukry** [comment]
I created the (repository)[https://github.com/pshoukry/shopify_issue] and confirmed the behavior. I just removed my api key.

**Comment by kevinhughes27** [comment]
Thanks I'll take a look as soon as I can!

**Comment by pshoukry** [comment]
Most welcome, let me know if i can help :)

**Comment by oleksandriuzikov** [comment]
This doesn't work for me.

```
config.redirect_uri = "<redirect_uri>"
```

It still adds `shop` parameter to the callback uri.

**Comment by pshoukry** [comment]
Yes it does this is just a work around:
1- remove it first time on install 
2- the app works fine after that.

**Comment by oleksandriuzikov** [comment]
@pshoukry thank you for the workaround but unfortunately this issue doesn't allow me to publish my app to the apps store because I cannot ask customers to remove parameters manually etc.

Can you please point me to the part of the code responsible for adding the `shop` param?

Thanks

**Comment by kevinhughes27** [comment]
Yeah I think that work around only works so you can develop the app. I'll
dig into this shortly and get some answers.

Some have suggested forcing your omniauth gem to an older version but I
haven't confirmed this fix yet.

**Comment by oleksandriuzikov** [comment]
@kevinhughes27 thank you! :beers: 

I've added this to my Gemfile:
```
gem 'omniauth-oauth2', '~> 1.2.0'
```

then updated with:
```
bundle update omniauth-oauth2
```

And the issue is gone.

**Comment by pshoukry** [comment]
Great thanks

**Comment by justinstern** [comment]
FWIW I was seeing this issue as well (extraneous `shop` param) until I configured both the `redirect_uri` and `callback_url` for the shopify omniauth provider, as shown in the [`shopify_provider.rb` template](https://github.com/Shopify/shopify_app/blob/4063b808f80c4e8cc89fb6d6378d50b9b377dab6/lib/generators/shopify_app/install/templates/shopify_provider.rb). Setting both of those to the same URL that I have configured as the Redirection URL in my partner account, stopped that `shop` param from being introduced.

With those two parameters configured, everything is working as expected with the following gems:

```
omniauth (1.2.2)
omniauth-oauth2 (1.4.0)
omniauth-shopify-oauth2 (1.1.10)
```

**Comment by kevinhughes27** [comment]
thanks @justinstern that is the fix - this happens by default in the recently release version of the gem (6.2.1) I noticed the issue repo only used 6.2.0

**Comment by tomtom87** [comment]
I tried installing from scratch from this repo to try and make my first shopify app and I am getting this exact same error of the `Oauth error invalid_request: The redirect_uri is missing or not whitelisted`

I have these gems that came with the package

```
omniauth 1.3.1
omnitauth-oauth2 1.4.0
omniauth-shopify-oauth2 (1.1.12)
```

Please assist I really want to develop an app for shopify!

**Comment by silaghidragos** [comment]
is there any news on this issues? i'm trying to develop an app for a client but the issues are still there

**Comment by tomtom87** [comment]
@silaghidragos same no one replied how to fix it yet so I moved onto other projects...

**Comment by kevinhughes27** [comment]
If you submit a repo that I can clone to reproduce the issue then I'll take a look.

**Comment by gautham20** [comment]
any idea on how to whitelist multiple redirect URLs? the wording the app settings screen suggests that this is possible. But I've the most common separators, but it doesn't seem to work.

**Comment by kevinhughes27** [comment]
I use newlines in the partner dashboard. Note that the setting in ruby code isn't required anymore so if you can still set this in ruby update your version of the gem.

**Comment by fabriziobertoglio1987** [comment]
Hello, 

I am having this error with Rails 5,  
* shopify_api (4.0.7)
* shopify_app (7.0.2)

I can not add a redirect uri both to the shopify_app.rb and to the omniauth.rb

**Comment by AmitPatel-BoTreeConsulting** [comment]
I am also facing same issue with Rails  5.1.3 and shopify_app (7.4.0)

I tried setting up `redirect_uri` in `shopify_app.rb` but getting error `undefined method `redirect_uri=' for #<ShopifyApp::Configuration:0x000000058697c8>`

The local setup works fine. Also the configuration works on non embedded app.

**Comment by kevinhughes27** [comment]
Please look at the configuration class. That option was removed several versions ago since it is not required.

**Comment by AmitPatel-BoTreeConsulting** [comment]
@kevinhughes27 thanks. I was trying the solution to fix the error.

![selection_922](https://user-images.githubusercontent.com/1289894/29855532-39dad8dc-8d6a-11e7-98ef-5ff1eb1a9b15.png)

**Comment by kevinhughes27** [comment]
You need to edit the urls in the partner dashboard not your app.

**Comment by AmitPatel-BoTreeConsulting** [comment]
@kevinhughes27 thanks. I have posted all the details in https://github.com/Shopify/shopify_app/issues/329#issuecomment-325888849
